examples/generative/speaker_recognition_using_cnn.py

```diff
@@ -0,0 +1,486 @@
+            if folder in [AUDIO_SUBFOLDER, NOISE_SUBFOLDER]:
+                continue
+            elif folder in NOISE_FOLDERS:
+                shutil.move(
+                    os.path.join(DATASET_ROOT, folder),
+                    os.path.join(DATASET_NOISE_PATH, folder),
+                )
+            else:
+                shutil.move(
+                    os.path.join(DATASET_ROOT, folder),
+                    os.path.join(DATASET_AUDIO_PATH, folder),
+                )
+
+# Let's generate the list of audio files and their corresponding label (speaker)
+(
+    file_paths,
+    labels,
+    class_names,
+) = tf.python.keras.preprocessing.dataset_utils.index_directory(
+    DATASET_AUDIO_PATH, "inferred", FILE_FORMATS, shuffle=True, seed=SHUFFLE_SEED
+)
+
+"""
+## Generating and Loading Noise
+"""
+
+# Let's first get the list of all noise files
+noise_paths, _, _ = tf.python.keras.preprocessing.dataset_utils.index_directory(
+    DATASET_NOISE_PATH, "inferred", FILE_FORMATS, shuffle=False
+)
+
+# - Read each of those files
+# - Resample it at the desired sampling rate
+# - Split it into chuncks of 1 second long each
+
+noise_stream = np.array([]).reshape(0, DESIRED_SR)
+
+for noise_path in noise_paths:
+    # Read and resample the noise audio file
+    noise, _ = librosa.load(noise_path, sr=DESIRED_SR)
+
+    # Split the audio into segments of 1 second long each
+    nb_samples = len(noise) // DESIRED_SR
+    arr = np.split(noise[: nb_samples * DESIRED_SR], nb_samples)
+
+    # Add the sample to the Noises array
+    noise_stream = np.concatenate((noise_stream, arr), axis=0)
+
+print(
+    "{} noise files were split into {} noise samples where each is {} sec. long".format(
+        len(noise_paths), noise_stream.shape[0], noise_stream.shape[1] // DESIRED_SR
+    )
+)
+
+# Create a random array of the same size as file_paths and with values
+# ranging from 0 to the number of noise stream samples that we have.
+# For every audio sample in file_paths, there will be a noise sample
+# identified by the value of noises at the same location as file_paths
+# This array will be shuffled at the end of every epoch in order to train
+# each sample with a wide range of noise samples.
+
+noises = np.random.randint(0, noise_stream.shape[0], len(file_paths))
+
+"""
+## Train / Validation Split
+"""
+
+valid_size = int(len(file_paths) * VALID_SPLIT)
+
+valid_file_paths, train_file_paths = file_paths[:valid_size], file_paths[valid_size:]
+valid_labels, train_labels = labels[:valid_size], labels[valid_size:]
+
+print("Training size: {}".format(len(train_file_paths)))
+print("Validation size: {}".format(len(valid_file_paths)))
+
+valid_noises, train_noises = noises[:valid_size], noises[valid_size:]
+
+print("Training noise size: {}".format(train_noises.shape[0]))
+print("Validation noise size: {}".format(valid_noises.shape[0]))
+
+"""
+## Data Generator
+"""
+
+
+class SpeechGenerator(keras.utils.Sequence):
```

Comment:
Consider using `tf.data` to load the data:

- start from file paths and their labels
- read_file https://www.tensorflow.org/api_docs/python/tf/io/read_file
- decode_wav https://www.tensorflow.org/api_docs/python/tf/audio/decode_wav
- fft https://www.tensorflow.org/api_docs/python/tf/signal/fft

Check out the code for `image_dataset_from_directory` for some inspiration.

This makes the example more scalable and more future-proof. It will also remove the dependency on scipy and librosa.

Reply:
All worked fine for reading audio waves using tf.audio.decode_wav for all the speakers’ samples in the dataset. In fact, all of those are sampled at sr=16000 
However for noise samples, those are at a different sampling rate (22050) this is why I used librosa which allows resampling while loading audio.
As an alternative, I believe we have 2 solutions:
- use tensorflow_io resample function (a requirement to have tfio)
- resample all audio files (6 files overall) using ffmpeg for example before using the tutorial.

Which one do you think is better for the case of this example?

Thanks!

Reply:
What would be your preference?

Since there's no need for the resampling to be online, I would think the second option (use of ffmpeg on the files on disk before making the dataset) is the simplest. But up to you!

Reply:
That’s exactly what I did today.
I resampled them on disk cause it’s easier + online resampling is not important for the purpose of the example.
Thanks!